### PR TITLE
SQL: Internal refactoring of operators as functions

### LINKED ISSUE #33975: SQL: Separate ProcessDefinition and ScriptTemplate from functions
When first introduced, `ProcessDefinition` and `ScriptTemplate` were part of `ScalarFunction` however the concept proved to me much more generic.
For example arithmetic operators were transformed into functions (despite not being one); the same situation needs to be applied with the rest of the operators, namely the comparison and some predicates (Like/RLike).
In fact, most (if not all) expressions require these as they can appear both on the invoking node and inside a script.

### PR DESCRIPTION
**Pull request by costin**
Sorry for the largish PR, the renaming and moving of some files had a broad effect (the PR actually adds only ~300 loc).

The main goal of this PR is to centralize and simplify the script generation between operators and functions which are currently decoupled. As part of this process most predicates (`<`, `<=`, etc...) were made `ScalarFunction` as their purpose and functionality is quite similar (see `%` and `MOD` functions).

~~In a similar vein, the script generation and processing was moved onto `Expression` itself as a general concern - it the future it might be tweaked a bit and~~was moved to the `NamedExpression`.
`ProcessDefinition` was renamed to `Pipe` to make the name shorter and a bit more clearer (a Pipe of processes) though it doesn't fully convey its purpose.
The advantage of moving this into the expression itself is that OO-ify the generation.
The script customization which was scattered across the code is now centralized into `ScriptWeaver` .
In the process, applied de-duplication over some function definitions, in particular String ones.

Lastly, this work started as part of the `NULL` support as every scripted operation, including operators (like `-`) need to be made null-safe. This essentially means having the same infrastructure across all pushed-down expressions.

Close #33975

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/ScriptWeaver.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                dataType());
+    }
+
+    default ScriptTemplate scriptWithField(FieldAttribute field) {
+        return new ScriptTemplate(formatScript("doc[{}].value"),
+                paramsBuilder().variable(field.name()).build(),
+                dataType());
+    }
+
+    default String formatScript(String template) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/ScriptWeaver.java]: maybe rename template -> script ?

**Review by matriv — APPROVED** [review]
LGTM. Nice stuff! Left two really minor comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryMathPipe.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import org.elasticsearch.xpack.sql.expression.function.scalar.math.BinaryMathProcessor.BinaryMathOperation;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.BinaryPipe;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+
+import java.util.Objects;
+
+/**
+ * Processor definition for math operations requiring two arguments.
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryMathPipe.java]: Not a processor definition anymore

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/BinaryStringNumericPipe.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.elasticsearch.xpack.sql.expression.function.scalar.string.BinaryStringNumericProcessor.BinaryStringNumericOperation;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.BinaryPipe;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+
+import java.util.Objects;
+
+/**
+ * Processor definition for String operations requiring one string and one numeric argument.
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/BinaryStringNumericPipe.java]: Not a processor definition anymore

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/BinaryStringStringPipe.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import org.elasticsearch.xpack.sql.expression.function.scalar.string.BinaryStringStringProcessor.BinaryStringStringOperation;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.BinaryPipe;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+
+import java.util.Objects;
+
+/**
+ * Processor definition for String operations requiring two string arguments.
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/BinaryStringStringPipe.java]: Not a processor definition anymore

**Review by astefan — APPROVED** [review]
Left some minor comments. LGTM

**Comment by costin** [comment]
Merged - thanks for the quick reviews!

**Pull request merged**